src/ops/verify_proof.rs

```diff
@@ -0,0 +1,141 @@
+                )
+                .map_err(|_| VMError::InvalidArguments)?;
+                let label =
+                    String::from_utf8(mem[label..label + label_len].to_vec())
+                        .map_err(|_| VMError::InvalidUtf8)?;
+
+                let pp = unsafe {
+                    PublicParameters::from_slice_unchecked(
+                        &rusk_profile::get_common_reference_string().map_err(
+                            |_| {
+                                VMError::ContractPanic(
+                                    "PubParams deser error".to_string(),
+                                )
+                            },
+                        )?,
+                    )
+                    .map_err(|_| VMError::InvalidArguments)?
+                };
+
+                match select_and_verify::<S>(&label, &pp, &vk, &pi_vec, &proof)
+                {
+                    Ok(()) => Ok(Some(RuntimeValue::I32(1))),
+                    Err(_) => Ok(Some(RuntimeValue::I32(0))),
+                }
+            })
+        } else {
+            Err(VMError::InvalidArguments)
+        }
+    }
+}
+
+fn decode_pub_inputs<S: Store>(
+    mem: &[u8],
+    pub_inp_len: usize,
+    pub_inp_addr: usize,
+) -> Result<Vec<PublicInput>, VMError<S>> {
+    mem[pub_inp_addr..pub_inp_addr + pub_inp_len]
+        .chunks(PublicInput::serialized_size())
+        .map(|chunk| {
+            PublicInput::from_bytes(chunk)
+                .map_err(|_| VMError::InvalidArguments)
+        })
+        .collect::<Result<Vec<PublicInput>, VMError<S>>>()
+}
+
+fn verify_proof<'a, S: Store>(
+    mut c: impl Circuit<'a>,
+    pp: &PublicParameters,
+    vk: &VerifierKey,
+    transcript_init: &'static [u8],
+    p_inp: &[PublicInput],
+    proof: &Proof,
+) -> Result<(), VMError<S>> {
+    c.verify_proof(pp, vk, transcript_init, proof, p_inp)
+        .map_err(|_| VMError::InvalidABICall)
+}
+
+#[rustfmt::skip]
+fn select_and_verify<'a, S:Store>(
+    label: &String,
+    pp: &PublicParameters,
+    vk: &VerifierKey,
+    p_inp: &[PublicInput],
+    proof: &Proof,
+) -> Result<(), VMError<S>> {
+    match label.as_str() {
+        "transfer-execute-1-0" => verify_proof(ExecuteCircuit::<17, 15>::default(), pp, vk, b"dusk" ,p_inp, proof),
+        "transfer-execute-1-1" => verify_proof(ExecuteCircuit::<17, 15>::default(), pp, vk, b"dusk" ,p_inp, proof),
+        "transfer-execute-1-2" => verify_proof(ExecuteCircuit::<17, 15>::default(), pp, vk, b"dusk" ,p_inp, proof),
+        "transfer-execute-2-0" => verify_proof(ExecuteCircuit::<17, 16>::default(), pp, vk, b"dusk" ,p_inp, proof),
+        "transfer-execute-2-1" => verify_proof(ExecuteCircuit::<17, 16>::default(), pp, vk, b"dusk" ,p_inp, proof),
+        "transfer-execute-2-2" => verify_proof(ExecuteCircuit::<17, 16>::default(), pp, vk, b"dusk" ,p_inp, proof),
+        "transfer-execute-3-0" => verify_proof(ExecuteCircuit::<17, 16>::default(), pp, vk, b"dusk" ,p_inp, proof),
```

Comment:
The capacity for circuits with 3 or more inputs should be `17`

Reply:
And it's 15 for the others? I took that from your transfer-circuits latest impl.